client/src/main/java/io/cloudsoft/winrm4j/client/WinRmClient.java

```diff
@@ -244,41 +252,71 @@ private int receiveCommand(String commandId, Writer out, Writer err) {
 
             //TODO use different instances of service http://cxf.apache.org/docs/developing-a-consumer.html#DevelopingaConsumer-SettingConnectionPropertieswithContexts
             setActionToContext((BindingProvider) winrm, "http://schemas.microsoft.com/wbem/wsman/1/windows/shell/Receive");
-            ReceiveResponse receiveResponse = winrm.receive(receive, RESOURCE_URI, MAX_ENVELOPER_SIZE, operationTimeout, locale, shellSelector);
-            List<StreamType> streams = receiveResponse.getStream();
-            for (StreamType s : streams) {
-                byte[] value = s.getValue();
-                if (value == null) continue;
-                if (out != null && "stdout".equals(s.getName())) {
-                    try {
-                        //TODO use passed locale?
-                        if (value.length > 0) {
-                            out.write(new String(value));
-                        }
-                        if (Boolean.TRUE.equals(s.isEnd())) {
-                            out.close();
-                        }
-                    } catch (IOException e) {
-                        throw new IllegalStateException(e);
+
+            try {
+                numberOfReceiveCalls++;
+                ReceiveResponse receiveResponse = winrm.receive(receive, RESOURCE_URI, MAX_ENVELOPER_SIZE, operationTimeout, locale, shellSelector);
+                getStreams(receiveResponse, out, err);
+
+                LOG.trace("receiveResponse: {}", receiveResponse);
+                CommandStateType state = receiveResponse.getCommandState();
+                if (COMMAND_STATE_DONE.equals(state.getState())) {
+                    return state.getExitCode().intValue();
+                }
+            } catch (SOAPFaultException soapFault) {
+                /**
+                 * If such Exception which has a code 2150858793 the client is expected to again trigger immediately a receive request.
+                 * https://msdn.microsoft.com/en-us/library/cc251676.aspx
+                 */
+
+                try {
+                    // e.getMessage().contains("The WS-Management service cannot complete the operation within the time specified in OperationTimeout")
+                    if (soapFault.getFault().getDetail().getFirstChild().getAttributes().getNamedItem("Code").getNodeValue().equals("2150858793")) {
+                        LOG.trace("winrm client {} received an operation response {}", this, soapFault);
+                    } else {
+                        throw soapFault;
                     }
+                } catch (NullPointerException e) {
```

Comment:
Personal preference for not using NPE exceptions for such logic flow. But no strong feelings - can leave it as-is.